syntax = "proto2";

// 相机畸变系数消息类型（支持k1-k6和p1-p2，p1/p2/k5/k6为可选字段）
message DistortCoeffs {
  double k1 = 1;   // 径向畸变系数1（必选）
  double k2 = 2;   // 径向畸变系数2（必选）
  double p1 = 5;   // 切向畸变系数1（可选）
  double p2 = 6;   // 切向畸变系数2（可选）
  double k3 = 3;   // 径向畸变系数3（必选）
  double k4 = 4;   // 径向畸变系数4（必选）
  double k5 = 7;   // 径向畸变系数5（可选）
  double k6 = 8;   // 径向畸变系数6（可选）
}

// 相机内参矩阵消息类型
message CameraMatrix {
  double fx = 1;  // 水平焦距
  double fy = 2;  // 垂直焦距
  double cx = 3;  // 图像中心x坐标
  double cy = 4;  // 图像中心y坐标
}

// 相机内参信息消息类型
message Intrinsics {
  string calibration_time = 1;       // 校准时间
  DistortCoeffs distort_coeffs = 2;  // 畸变系数（含可选扩展）
  CameraMatrix camera_matrix = 3;    // 内参矩阵
  string serial_no = 4;              // 相机序列号
  string calibration_engineer = 5;   // 校准工程师
  double rms = 6;                    // 均方根误差
}

// 相机配置消息类型
message CameraConfig {
  Intrinsics intrinsics = 1;                  // 相机内参
  int32 max_fps = 2;                          // 最大帧率
  double mid_row_time_since_trigger = 3;      // 触发后中间行时间
  double rolling_elapse_time = 4;             // 滚动曝光时间
  string camera_vendor = 5;                   // 相机厂商
  string serial_no = 6;                       // 相机序列号
}